R/functions-eager.R

```diff
@@ -1,25 +1,223 @@
+          )$
+            drop(paste0(common_cols, "_PL_CONCAT_RIGHT"))
+        },
+        accumulate = TRUE
+      )$sort(common_cols)
+
+      if (is_polars_df(first)) {
+        out <- lf$collect()
+      } else {
+        out <- lf
+      }
+      return(out)
+    }
+
+    out <- if (is_polars_df(first)) {
+      how <- arg_match0(
+        how,
+        values = c("vertical", "vertical_relaxed", "diagonal", "diagonal_relaxed", "horizontal")
+      )
+      switch(how,
+        "vertical" = {
+          items |>
+            lapply(\(x) x$`_df`) |>
+            concat_df() |>
+            wrap()
+        },
+        "vertical_relaxed" = {
+          res <- items |>
+            lapply(\(x) x$lazy()$`_ldf`) |>
+            concat_lf(
+              rechunk = rechunk,
+              parallel = parallel,
+              to_supertypes = TRUE
+            ) |>
+            wrap()
+          res$collect(no_optimization = TRUE)
+        },
+        "diagonal" = {
+          items |>
+            lapply(\(x) x$`_df`) |>
+            concat_df_diagonal() |>
+            wrap()
+        },
+        "diagonal_relaxed" = {
+          res <- items |>
+            lapply(\(x) x$lazy()$`_ldf`) |>
+            concat_lf_diagonal(
+              rechunk = rechunk,
+              parallel = parallel,
+              to_supertypes = TRUE
+            ) |>
+            wrap()
+          res$collect(no_optimization = TRUE)
+        },
+        "horizontal" = {
+          items |>
+            lapply(\(x) x$`_df`) |>
+            concat_df_horizontal() |>
+            wrap()
+        }
+      )
+    } else if (is_polars_lf(first)) {
+      switch(how,
+        "vertical" = ,
+        "vertical_relaxed" = {
+          items |>
+            lapply(\(x) x$`_ldf`) |>
+            concat_lf(
+              rechunk = rechunk,
+              parallel = parallel,
+              to_supertypes = endsWith(how, "relaxed")
+            ) |>
+            wrap()
+        },
+        "diagonal" = ,
+        "diagonal_relaxed" = {
+          items |>
+            lapply(\(x) x$`_ldf`) |>
+            concat_lf_diagonal(
+              rechunk = rechunk,
+              parallel = parallel,
+              to_supertypes = endsWith(how, "relaxed")
+            ) |>
+            wrap()
+        },
+        "horizontal" = {
+          items |>
+            lapply(\(x) x$`_ldf`) |>
+            concat_lf_horizontal(parallel = parallel) |>
+            wrap()
+        }
+      )
+    } else if (is_polars_series(first)) {
+      if (how == "vertical") {
+        items |>
+          lapply(\(x) x$`_s`) |>
+          concat_series() |>
+          wrap()
+      } else {
+        abort("Series only supports 'vertical' concat strategy.")
+      }
+    } else if (is_polars_expr(first)) {
+      return(
+        items |>
+          lapply(\(x) x$`_rexpr`) |>
+          concat_expr(rechunk = rechunk) |>
+          wrap()
+      )
+    } else {
```

Comment:
This is indeed the same as Python Polars, but it looks rather odd not to check `how`......

Reply:
Yes, I think it's missing in py-polars. I'll add more checks tomorrow (I'm also unsure of what the accepted values are for Expr inputs).

Reply:
https://github.com/pola-rs/polars/issues/19813

Reply:
I have removed the code and tests related to Expr following the answer in the issue above.